source/agora/test/ManyValidators.d

```diff
@@ -50,201 +44,62 @@ unittest
-        .sign();
-    txs ~= spendable[7].split(WK.Keys.Genesis.address.repeat(8)).sign();
-    txs.each!(tx => nodes[0].putTransaction(tx));
-    // block 17
-    network.expectBlock(expected_block++);
-
-    // freeze builders
-    auto freezable = txs[$ - 2]  // contains 16 payment UTXOs
-        .outputs.length.iota
-        .takeExactly(conf.outsider_validators + GenesisValidators)  // there might be more UTXOs
-        .map!(idx => TxBuilder(txs[$ - 2], cast(uint)idx))
-        .array;
+    // generate 18 blocks, 2 short of the enrollments expiring.
+    network.generateBlocks(Height(GenesisValidatorCycle - 2));
 
-    // create 16 freeze TXs
-    auto freeze_txs = freezable
-        .enumerate
-        .map!(pair => pair.value.refund(WK.Keys[pair.index].address)
-            .sign(TxType.Freeze))
+    const keys = network.nodes.map!(node => node.client.getPublicKey())
+        .dropExactly(GenesisValidators).takeExactly(conf.outsider_validators)
         .array;
 
-    // block 18
-    freeze_txs[0 .. 8].each!(tx => nodes[0].putTransaction(tx));
-    network.expectBlock(expected_block++);
+    // prepare frozen outputs for outsider validators to enroll
+    genesisSpendable().dropExactly(1).takeExactly(1)
+        .map!(txb => txb.split(keys).sign(TxType.Freeze))
+        .each!(tx => nodes[0].putTransaction(tx));
 
     // block 19
-    freeze_txs[8 .. 16].each!(tx => nodes[0].putTransaction(tx));
-    network.expectBlock(expected_block++);
-
-    // now we re-enroll existing validators (extension),
-    // and enroll 10 new validators.
-    foreach (node; nodes)
-    {
-        Enrollment enroll = node.createEnrollmentData();
-        node.enrollValidator(enroll);
-
-        // check enrollment
-        nodes.each!(n =>
-            retryFor(n.getEnrollment(enroll.utxo_key) == enroll, 5.seconds));
-    }
-
-    // at block height 20 the validator set changes
-    txs = txs[$ - 1]  // take those 8 UTXOs from #L184
-            .outputs.length.iota.map!(idx => TxBuilder(txs[$ - 1], cast(uint)idx))
-            .takeExactly(8)  // there might be more than 8
-            .map!(txb => txb.refund(WK.Keys.Genesis.address).sign()).array;
-        txs.each!(tx => nodes[0].putTransaction(tx));
-
-    // Block 20
-    network.expectBlock(expected_block++);
-
-    // sanity check
+    network.generateBlocks(Height(GenesisValidatorCycle - 1));
+
+    // wait for other nodes to get to same block height
+    nodes.drop(GenesisValidators).enumerate.each!((idx, node) =>
+        retryFor(node.getBlockHeight() == GenesisValidatorCycle - 1, 2.seconds,
+            format!"Expected block height %s but outsider %s has height %s."
+                (GenesisValidatorCycle - 1, idx, node.getBlockHeight())));
+
+    // Now we enroll new validators and re-enroll the original validators
+    iota(0, validators).each!(idx => network.enroll(idx));
+
+    network.generateBlocks(Height(GenesisValidatorCycle), Height(0),
+        iota(nodes.length).array);
```

Comment:
See now you're forcing the call site to use arrays. Why not just make `generateBlocks` take a range? Then it can accept either an input range or an array.

Reply:
Yes that would be better. I will update.

Reply:
Updated to use range again.